lib/property_sets/delegator.rb

```diff
@@ -40,9 +40,15 @@ def delegate_to_property_set(setname, mappings)
           alias_method "#{old_attr}_before_type_cast", old_attr
           define_method("#{old_attr}?") { send(setname).send("#{new_attr}?") }
           define_method("#{old_attr}=") do |value|
-            attribute_will_change!(old_attr) if old_attr != value && !defined?(super)
+            if send(old_attr) != value
+              send("#{old_attr}_will_change!")
+            end
             send(setname).send("#{new_attr}=", value)
-            super(value) if defined?(super)
+            super(value) if defined?(super) # Rails 4 does not define this
```

Comment:
Which method is not defined by Rails 4? It looks like you mean `#attr=`, but attribute setters surely were defined in Rails 4 (and earlier).

Reply:
When run _without_ `if defined?(sure)` the specs under rails 4 will raise errors like:
```
NoMethodError:
        super: no superclass method `old=' for #<Account:0x00007fa5ffacaab0>
        Did you mean?  old
                       old?
```

Reply:
Isn’t that a bit strange? Where are the setter methods defined in a Rails 4 model?

Reply:
Well it's interesting.
In rails 4:
```
self.methods.grep(/old/)
[:old, :old?, :old_will_change!, :old=, :old_changed?, :old_before_type_cast]
self.methods.grep(/old/).map{|name| puts method(name).source_location }
.../property_sets/lib/property_sets/delegator.rb,35
.../property_sets/lib/property_sets/delegator.rb,41
.../property_sets/lib/property_sets/delegator.rb,51
.../property_sets/lib/property_sets/delegator.rb,42
.../property_sets/lib/property_sets/delegator.rb,55
.../property_sets/lib/property_sets/delegator.rb,35
```
And in rails 5 there's more, that makes me think we're missing some features
```
self.methods.grep(/old/)
[:old, :old=, :old_before_type_cast, :old_changed?, :old?, :old_will_change!, :old_came_from_user?, :old_change, :old_was, :old_previously_changed?, :old_previous_change, :restore_old!]
self.methods.grep(/old/).map{|name| puts method(name).source_location }
.../property_sets/lib/property_sets/delegator.rb,35
.../property_sets/lib/property_sets/delegator.rb,42
.../property_sets/lib/property_sets/delegator.rb,35
.../property_sets/lib/property_sets/delegator.rb,55
.../property_sets/lib/property_sets/delegator.rb,41
.../property_sets/lib/property_sets/delegator.rb,51
../.gem/ruby/2.6.5/gems/activemodel-5.0.7.2/lib/active_model/attribute_methods.rb,384
../.gem/ruby/2.6.5/gems/activemodel-5.0.7.2/lib/active_model/attribute_methods.rb,384
../.gem/ruby/2.6.5/gems/activemodel-5.0.7.2/lib/active_model/attribute_methods.rb,384
../.gem/ruby/2.6.5/gems/activemodel-5.0.7.2/lib/active_model/attribute_methods.rb,384
../.gem/ruby/2.6.5/gems/activemodel-5.0.7.2/lib/active_model/attribute_methods.rb,384
../.gem/ruby/2.6.5/gems/activemodel-5.0.7.2/lib/active_model/attribute_methods.rb,384
```
Rails 5.1
```
self.methods.grep(/old/)
[:old?, :old_will_change!, :old, :old=, :old_changed?, :old_before_type_cast, :old_came_from_user?, :old_was, :old_previously_changed?, :old_previous_change, :restore_old!, :old_change, :will_save_change_to_old?, :saved_change_to_old, :old_before_last_save, :old_change_to_be_saved, :saved_change_to_old?, :old_in_database]
# with similar output as 5.0 for source location
```
Still  a different set in Rails 5.2
```
 self.methods.grep(/old/)
[:old, :old?, :old_will_change!, :old=, :old_changed?, :old_before_type_cast, :old_change, :old_came_from_user?, :old_previously_changed?, :old_was, :old_previous_change, :will_save_change_to_old?, :old_change_to_be_saved, :old_in_database, :restore_old!, :saved_change_to_old?, :saved_change_to_old, :old_before_last_save]
# same locations as 5.1 and 5.0
```

Reply:
This may have to do with load order and some portion of rails checking to see if a setter already exists before defining it. I can keep digging, but given that this is to support rails 4, this check will likely go away soon and is probably not worth pursuing just for the comment given that the released code already has this check